### PR TITLE
Mid cycle report - National level

### PR DESCRIPTION
**Pull request by JR-G**
## Context

Around the 22nd May, we’ll be sending out correspondence to all providers about their performance up until the mid point of this cycle.

As part of this, we’ll be sharing the national performance for things like:

- Reject by default rate
- acceptances this cycle versus last cycle

This report will be publicly available.

## Changes proposed in this pull request

We've got a mid cycle report!

(https://www.apply-for-teacher-training.service.gov.uk/publications/mid-cycle-report)

image

----

_Mid cycle report released
Provider performance displayed
Teacher training insights_

## Guidance to review

The dates `11 October 2022 to 3 April 2023` appear quite often but are hardcoded for now. It's a priority to get this out in time, but I do want to avoid introducing too much tech debt, so I think this does require some thought as to how best we can future proof it.

This data has also come straight from the D&I team, and has been QA'd their end. I'll upload the data to the review app, so we can test.

## Link to Trello card

https://trello.com/c/QkZAj0t4

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,141 @@
+<% if @national_data.nil? %>
+  <p class="govuk-body">This report is not ready to view.</p>
+<% else %>
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <h1 class="govuk-heading-l">
+        National recruitment performance for the 2022 to 2023 recruitment cycle
+      </h1>
+      <p class="govuk-body">These statistics show the recruitment performance of all initial teacher training (ITT) providers in England from 11 October 2022 to 3 April 2023.</p>
+      <p class="govuk-body">If your organisation is an ITT provider, and you use the Manage teacher training applications (Manage) service, you can view your own organisation's recruitment performance by signing in to <%= govuk_link_to('Manage', provider_interface_sign_in_path) %>.</p>
```
Comment (MaeveRoseveare) [inline app/views/publications/mid_cycle_report/show.html.erb:10]:
```suggestion
      <p class="govuk-body">If your organisation is an ITT provider, and you use the Manage teacher training applications (Manage) service, you can view your own organisation's recruitment performance by <%= govuk_link_to('signing in to Manage', provider_interface_sign_in_path) %>.</p>
```

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      <h2 class="govuk-heading-m">How the data in this report compares to other data we publish</h2>
+      <p class="govuk-body">The page showing <%= govuk_link_to('monthly statistics for courses starting in the 2023 to 2024 academic year', publications_monthly_report_itt_path(RecruitmentCycle.current_year)) %>, calculates data in the same way UCAS used to.</p>
+      <p class="govuk-body">This page about national recruitment performance, calculates data in a new, improved way. For this reason, you should not compare it with the monthly statistics page. </p>
+      <h2 class="govuk-heading-m">Candidates who've submitted applications</h2>
```
Comment (MaeveRoseveare) [inline app/views/publications/mid_cycle_report/show.html.erb:14]: For this table, from Katie's doc we should only show overall data, so just total for Secondary and total for Primary, we don't need to split by subject. See provider report:

Screenshot 2023-05-12 at 12 57 32

Comment (frankieroberto) [inline app/views/publications/mid_cycle_report/show.html.erb:14]:
```suggestion
      <h2 class="govuk-heading-m">Candidates who’ve submitted applications</h2>
```

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      <p class="govuk-body">This table shows the number of candidates who’ve submitted at least one application to an ITT provider from 10 October 2022 to 3 April 2023.</p>
```
Comment (frankieroberto) [inline app/views/publications/mid_cycle_report/show.html.erb:15]:
```suggestion
      <p class="govuk-body">This table shows the number of candidates who submitted at least one application to an ITT provider between 10 October 2022 to 3 April 2023.</p>
```

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      <table class="govuk-table govuk-!-margin-bottom-9">
+        <%= render 'table_body', model: @national_data, first_column: 'Candidates submitted', second_column: 'Candidates submitted (previous cycle)', third_column: 'Candidates submitted change', include_total: true %>
+      </table>
+    </div>
+  </div>
+
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <h2 class="govuk-heading-m">Number of applications</h2>
+      <p class="govuk-body">This table shows the total number of applications received by all ITT providers from 11 October 2022 to 3 April 2023.</p>
```
Comment (frankieroberto) [inline app/views/publications/mid_cycle_report/show.html.erb:29]: @MaeveRoseveare could we drop "This table shows" from all the intro paragraphs maybe?

```suggestion
      <p class="govuk-body">The total number of applications received by all ITT providers from 11 October 2022 to 3 April 2023.</p>
```

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    <div class="govuk-grid-column-full">
+      <table class="govuk-table govuk-!-margin-bottom-9">
+        <%= render 'table_body', model: @national_data, first_column: 'Candidates submitted', second_column: 'Candidates submitted (previous cycle)', third_column: 'Candidates submitted change', include_total: true %>
+      </table>
+    </div>
+  </div>
+
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <h2 class="govuk-heading-m">Number of applications</h2>
```
Comment (frankieroberto) [inline app/views/publications/mid_cycle_report/show.html.erb:28]:
```suggestion
      <h2 class="govuk-heading-m">Applications submitted</h2>
```

File: app/views/publications/mid_cycle_report/_table_body.html.erb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+<thead class="govuk-table__head">
+  <tr class="govuk-table__row">
+    <th scope="col" class="govuk-table__header">Level</th>
+    <th scope="col" class="govuk-table__header">This cycle</th>
+    <th scope="col" class="govuk-table__header">Last cycle</th>
+    <th scope="col" class="govuk-table__header">Percentage change</th>
```
Comment (frankieroberto) [inline app/views/publications/mid_cycle_report/_table_body.html.erb:6]: Don't think we need to include "Percentage" in the header, as each of the cells have their own percentage sign.

```suggestion
    <th scope="col" class="govuk-table__header">Change</th>
```

File: app/views/publications/mid_cycle_report/_table_body.html.erb [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  </tr>
+</thead>
+<tbody class="govuk-table__body">
+  <% model.each do |row| %>
+    <tr class="govuk-table__row">
+      <th scope="row" class="govuk-table__header govuk-!-font-weight-regular"><%= row['tad_subject'] %></th>
+      <td class="govuk-table__cell"><%= row[first_column] %></td>
+      <td class="govuk-table__cell"><%= row[second_column] %></td>
+      <td class="govuk-table__cell"><%= row[third_column] %></td>
```
Comment (frankieroberto) [inline app/views/publications/mid_cycle_report/_table_body.html.erb:15]: Adding this extra class makes the numbers right-aligned and using a fixed-width font. This makes the numbers easier to compare visually.

```suggestion
      <td class="govuk-table__cell govuk-table__cell--numeric"><%= row[first_column] %></td>
      <td class="govuk-table__cell govuk-table__cell--numeric"><%= row[second_column] %></td>
      <td class="govuk-table__cell govuk-table__cell--numeric"><%= row[third_column] %></td>
```

File: app/views/publications/mid_cycle_report/_table_body.html.erb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    </tr>
+  <% end %>
+</tbody>
+<% if include_total %>
+  <tfoot>
+    <tr class="govuk-table__row">
+      <th scope="row" class="govuk-table__header">Total</th>
+      <th scope="row" class="govuk-table__header"><%= number_with_delimiter(model.map { |row| row[first_column].delete(',').to_i }.sum, delimiter: ',') %></th>
+      <th scope="row" class="govuk-table__header"><%= number_with_delimiter(model.map { |row| row[second_column].delete(',').to_i }.sum, delimiter: ',') %></th>
+      <th scope="row" class="govuk-table__header">-</th>
```
Comment (frankieroberto) [inline app/views/publications/mid_cycle_report/_table_body.html.erb:25]: I think semantically these should be `td` cells not `th`, as they contain data not headings. I've added a separate override class to make them bold again.

```suggestion
      <td class="govuk-table__cell govuk-table__cell--numeric govuk-!-font-weight-bold"><%= number_with_delimiter(model.map { |row| row[first_column].delete(',').to_i }.sum, delimiter: ',') %></td>
      <td class="govuk-table__cell govuk-table__cell--numeric govuk-!-font-weight-bold"><%= number_with_delimiter(model.map { |row| row[second_column].delete(',').to_i }.sum, delimiter: ',') %></td>
      <td class="govuk-table__cell govuk-table__cell--numeric govuk-!-font-weight-bold">-</td>
```

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+      <h2 class="govuk-heading-m">How the data in this report compares to other data we publish</h2>
+      <p class="govuk-body">The page showing <%= govuk_link_to('monthly statistics for courses starting in the 2023 to 2024 academic year', publications_monthly_report_itt_path(RecruitmentCycle.current_year)) %>, calculates data in the same way UCAS used to.</p>
+      <p class="govuk-body">This page about national recruitment performance, calculates data in a new, improved way. For this reason, you should not compare it with the monthly statistics page. </p>
+      <h2 class="govuk-heading-m">Candidates who've submitted applications</h2>
+      <p class="govuk-body">This table shows the number of candidates who’ve submitted at least one application to an ITT provider from 10 October 2022 to 3 April 2023.</p>
+    </div>
+  </div>
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-full">
+      <table class="govuk-table govuk-!-margin-bottom-9">
```
Comment (frankieroberto) [inline app/views/publications/mid_cycle_report/show.html.erb:20]: Move the `<table>` and `</table>` to the partial maybe? Feels weird to separate the table header, body and container from its footer? Then rename the partial `_table`. 😄

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+      <h2 class="govuk-heading-m">Deferrals</h2>
+      <p class="govuk-body">This table shows the number of deferrals that have happened so far this cycle (from 11 October 2022 to 3 April 2023). It also shows deferrals from last cycle, where a candidate deferred their offer into this cycle.</p>
+    </div>
+  </div>
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-full">
+      <table class="govuk-table govuk-!-margin-bottom-9">
+        <thead class="govuk-table__head">
+          <tr class="govuk-table__row">
+            <th scope="col" class="govuk-table__header">Dererrals</th>
```
Comment (frankieroberto) [inline app/views/publications/mid_cycle_report/show.html.erb:80]: Typo 😉 
```suggestion
            <th scope="col" class="govuk-table__header"> Deferrals </th>
```

**Review by frankieroberto — DISMISSED** [review]
Looking good!

A few more things:

1. Can we change the "Level" column heading to "Subject"? (most rows are subjects)
2. "Primary" row should appear first, then subjects in alphabetical order, except with "Others" as the last row.
3. Can we add a "Secondary" row, appearing after Primary, which contains a total across all secondary subjects? Then the secondary subjects could be slightly indented?
4. The total rows should have percentage change between cycles rather than a `-` so that people can see the overall change.

**Review by MaeveRoseveare — CHANGES_REQUESTED** [review]

File: app/views/publications/mid_cycle_report/show.html.erb (hunk 4 above)
Comment (MaeveRoseveare) [reply app/views/publications/mid_cycle_report/show.html.erb:29]: Sounds good!

**Comment by MaeveRoseveare** [comment]
For subject can we write the double barrels is lower case, so:
Modern foreign languages
Religions education
Business studies
Art and design
Design and technology

File: app/views/publications/mid_cycle_report/show.html.erb (hunk 10 above)
Comment (JR-G) [reply app/views/publications/mid_cycle_report/show.html.erb:80]: 🙈

Comment (JR-G) [reply app/views/publications/mid_cycle_report/show.html.erb:80]: _derer rals_

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    <div class="govuk-grid-column-full">
+      <table class="govuk-table govuk-!-margin-bottom-9">
+        <%= render 'table', model: @national_data, first_column: 'Candidates submitted', second_column: 'Candidates submitted (previous cycle)', third_column: 'Candidates submitted change', include_total: true, top_level_only: true %>
+      </table>
+    </div>
+  </div>
+
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <h2 class="govuk-heading-m" id="number-applications">Applications submitted</h2>
```
Comment (MaeveRoseveare) [inline app/views/publications/mid_cycle_report/show.html.erb:85]:
```suggestion
      <h2 class="govuk-heading-m" id="number-applications">Number of applications</h2>
```

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+  </div>
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-full">
+        <%= render 'table', model: @national_data, first_column: 'RBD (this cycle)', second_column: 'RBD (previous cycle)', third_column: 'RBD change', include_total: true, top_level_only: false %>
+    </div>
+  </div>
+
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <h2 class="govuk-heading-m" id="proportion">Applications rejected</h2>
```
Comment (MaeveRoseveare) [inline app/views/publications/mid_cycle_report/show.html.erb:173]:
```suggestion
      <h2 class="govuk-heading-m" id="proportion">Proportion of applications rejected</h2>
```

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+      <p class="govuk-body">The page showing <%= govuk_link_to('monthly statistics for courses starting in the 2023 to 2024 academic year', publications_monthly_report_itt_path(RecruitmentCycle.current_year)) %>, calculates data using the same method UCAS used to. This method counts a candidate’s furthest application. For example, if an application moves from ‘recruited’ to ‘deferred’ the recruited count will go down, and the deferred count will go up. This method does not easily show where a candidate ends up over time.</p>
+      <p class="govuk-body">This page about national recruitment performance, calculates data in a new way by counting all submitted applications, regardless of their outcome. It also counts all offers regardless of if the offer is accepted, declined or waiting for a decision.</p>
+      <p class="govuk-body">With each page using a different method, you should not compare this page with the monthly statistics page.</p>
+      <div class="govuk-grid-row">
+        <div class="govuk-grid-column-two-thirds">
+          <h2 class="govuk-body govuk-!-font-size-16 govuk-!-margin-bottom-1 govuk-!-margin-top-5">Contents</h2>
```
Comment (MaeveRoseveare) [inline app/views/publications/mid_cycle_report/show.html.erb:17]: Create the same change as the provider report, where we move the contents list to before the header 'How the data in this report compares...' and make this header number 1 in the list. And add the same corresponding numbers to each H2.
https://ukgovernmentdfe.slack.com/archives/C04U0N9NYFP/p1684164132924849?thread_ts=1684159801.971229&cid=C04U0N9NYFP

**Review by MaeveRoseveare — CHANGES_REQUESTED** [review]

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+      <p class="govuk-body">The page showing <%= govuk_link_to('monthly statistics for courses starting in the 2023 to 2024 academic year', publications_monthly_report_itt_path(RecruitmentCycle.current_year)) %>, calculates data using the same method UCAS used to. This method counts a candidate’s furthest application. For example, if an application moves from ‘recruited’ to ‘deferred’ the recruited count will go down, and the deferred count will go up. This method does not easily show where a candidate ends up over time.</p>
+      <p class="govuk-body">This page about national recruitment performance, calculates data in a new way by counting all submitted applications, regardless of their outcome. It also counts all offers regardless of if the offer is accepted, declined or waiting for a decision.</p>
+      <p class="govuk-body">With each page using a different method, you should not compare this page with the monthly statistics page.</p>
+    </div>
+  </div>
+
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <h2 class="govuk-heading-m" id="submitted">2. Candidates who’ve submitted applications</h2>
```
Comment (MaeveRoseveare) [inline app/views/publications/mid_cycle_report/show.html.erb:102]: I'm hoping this is the last content change! But for this table, where we only show Primary or Secondary. Can we change the table header from 'Subject' back to '**Level**'? As I don't think subject makes sense for this one.

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+            </li>
+          </ol>
+        </div>
+      </div>
+    </div>
+  </div>
+
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <h2 class="govuk-heading-m" id="data">How this data compares to other data we publish</h2>
```
Comment (MaeveRoseveare) [inline app/views/publications/mid_cycle_report/show.html.erb:93]:
```suggestion
      <h2 class="govuk-heading-m" id="data">1. How this data compares to other data we publish</h2>
```

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-full">
+      <%= render 'table', model: @national_data, first_column: 'Applications submitted (this cycle)', second_column: 'Applications submitted (previous cycle)', third_column: 'Applications submitted change', include_total: true, top_level_only: false, subject_table: true %>
+    </div>
+  </div>
+
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <h2 class="govuk-heading-m" id="offer">4. Applications with an offer</h2>
+      <p class="govuk-body">The number of applications with an offer from an ITT training provider from 11 October 2022 to 3 April 2023.</p>
```
Comment (MaeveRoseveare) [inline app/views/publications/mid_cycle_report/show.html.erb:129]: **Replace content with:**
The number of applications with an offer made to a candidate between 11 October 2022 and 3 April 2023.  

The table includes applications: 

- with offers withdrawn by the provider 
- withdrawn by candidates 
- deferred until next cycle 

The table does not include deferrals from a previous cycle and then reconfirmed by the provider this cycle.

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-full">
+      <%= render 'table', model: @national_data, first_column: 'Offers made (this cycle)', second_column: 'Offers made (previous cycle)', third_column: 'Offers made change', include_total: true, top_level_only: false, subject_table: true %>
+    </div>
+  </div>
+
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <h2 class="govuk-heading-m" id="accepted">5. Offers accepted</h2>
+      <p class="govuk-body">The number of offers that have been accepted from 11 October 2022 to 3 April 2023.</p>
```
Comment (MaeveRoseveare) [inline app/views/publications/mid_cycle_report/show.html.erb:141]: **Replace content with:**
The number of candidates who accepted an offer so far this cycle.  

The table includes candidates who accepted and then deferred an offer in a previous cycle, and the provider has reconfirmed the same offer this cycle. 

The table does not include applications: 

- with offers withdrawn by the provider 
- withdrawn by candidates 
- deferred until the next cycle

File: app/views/publications/mid_cycle_report/show.html.erb [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-full">
+      <%= render 'table', model: @national_data, first_column: 'Acceptances (this cycle)', second_column: 'Acceptance (previous cycle)', third_column: 'Acceptances change', include_total: true, top_level_only: false, subject_table: true %>
+    </div>
+  </div>
+
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+      <h2 class="govuk-heading-m" id="deferrals">6. Deferrals</h2>
+      <P class="govuk-body">The number of deferrals that have happened so far this cycle (from 11 October 2022 to 3 April 2023). It also shows deferrals from last cycle, where a candidate deferred their offer into this cycle.</p>
```
Comment (MaeveRoseveare) [inline app/views/publications/mid_cycle_report/show.html.erb:154]: **Replace content with:**
The number of deferrals so far this cycle from 11 October 2022 to 3 April 2023. 

The table includes candidates who: 

- accepted and deferred an offer in a previous cycle, and the provider has reconfirmed the same offer this cycle 
- accepted an offer this cycle, but deferred the same offer until the next cycle 

The table does not include withdrawn applications.

**Pull request merged**